### DISCUSSION TITLE
Concealing wikilink with note's title

### DISCUSSION THREAD

**Discussion by kirasok** [MeanderingProgrammer/render-markdown.nvim#228] [Show and tell]
I use `render-markdown.nvim` with [zk-nvim](https://github.com/zk-org/zk-nvim) for my notes. `zk-nvim` provides title of a note in diagnostics. I wanted to conceal wikilink with that title and made [a fork](https://github.com/kirasok/render-markdown.nvim) that does it.

```markdown
# index

- [[rigo|first note]]
- [second note](zzl2)
- [[f8fu]]
```

<details>
<summary>without fork</summary>
<img src="https://github.com/user-attachments/assets/c5293f83-ddfa-4b4c-aa54-aad0fba8ac8b">
</details>

<details>
<summary>with fork</summary>
<img src="https://github.com/user-attachments/assets/c670de08-4252-4221-bf90-7072a8ae643c">
</details>

**Comment by MeanderingProgrammer** [comment]
Populating content based on information from diagnostic is an interesting idea, thanks for sharing!

Adding the changes needed here as a reference in case the fork goes away:

- Add `DiagnosticChanged` as a change event
- Add `fetch_from_diagnostics` as a configuration option
- Add method to get information from diagnostics based on node, for example:

```lua
---@param node render.md.Node
---@return string?
function Context:get_diagnostic(node)
    local row, col = node.start_row, node.start_col
    local diagnostics = vim.diagnostic.get(self.buf, {
        severity = vim.diagnostic.severity.HINT,
    })
    for _, diagnostic in pairs(diagnostics) do
        if diagnostic.col == col and diagnostic.lnum == row then
            return diagnostic.message
        end
    end
    return nil
end
```

**Comment by mcDevnagh** [comment]
look like this fork has some merge conflicts. 
I created a new fork, updating the code from @kirasok and incorporating the change @MeanderingProgrammer mentioned: moving the `get_diagnostic` function to `Context`: https://github.com/MeanderingProgrammer/render-markdown.nvim/pull/345

@MeanderingProgrammer Do you think we can get this merged to upstream, so we don't run into an outdated fork again?

**Comment by MeanderingProgrammer** [comment]
I've implemented a version of this here: https://github.com/MeanderingProgrammer/render-markdown.nvim/commit/0df6719abc3b547bc5b7111a750d8d7e035a7234, which takes inspiration from your PR. Unless more use cases crop up I wanted to keep the diagnostics specific logic out of the repo, but still enable the feature. 

After updating you'll need to set something like this as your configuration:


```lua
require('render-markdown').setup({
    change_events = { 'DiagnosticChanged' },
    link = {
        wiki = {
            body = function(ctx)
                local diagnostics = vim.diagnostic.get(ctx.buf, {
                    lnum = ctx.row,
                    severity = vim.diagnostic.severity.HINT,
                })
                for _, diagnostic in ipairs(diagnostics) do
                    if diagnostic.col == ctx.start_col and diagnostic.source == 'zk' then
                        return diagnostic.message
                    end
                end
                return nil
            end,
        },
    },
})
```

**Reply by mcDevnagh** [reply]
Thank you so much!
I do have 2 more issues:
1) I'd like the alias to display when one exists, `[[destination|alias]]`, so I created this PR #350, and it worked as expected with the following config: 
```lua
require('render-markdown').setup({
    link = {
        wiki = {
            body = function(ctx)
                if ctx.alias then
                    return ctx.alias
                end
                local diagnostics = vim.diagnostic.get(ctx.buf, {
                    lnum = ctx.row,
                    severity = vim.diagnostic.severity.HINT,
                })
                for _, diagnostic in ipairs(diagnostics) do
                    if diagnostic.source == 'zk' then
                        return diagnostic.message
                    end
                end
                return nil
            end,
        },
    },
})
```

2) Underlying Highlight Groups of the text seems to be ignored:
In this example, I want the link to appear as if it's a header. Is that possible?
![image](https://github.com/user-attachments/assets/cbcf236b-7609-47fd-89c1-774b9283e90f)
```md
# [[b48069e7]]
# [Random Encounters](rules/6-game-master/b48069e7)
- Reaction Check
- Treasure
```

Thanks, again

**Comment by MeanderingProgrammer** [comment]
> Underlying Highlight Groups of the text seems to be ignored:
> In this example, I want the link to appear as if it's a header. Is that possible?

Rather than returning the alias when `ctx.alias` exists I would early return `nil`. 

The behavior of rendering `body` needs to be able to handle the user returning anything so we can't make assumptions that they're providing a substring of the `wikilink`. 

When the returned value is `nil` we fallback to the default rendering behavior, which doesn't `inline` virtual text to render aliases, instead it only conceals the destination and keeps the alias as is, which is why the highlight groups look more in place.

The difference is when body returns a `string` we run:

```lua
self.marks:add_over('link', self.node, {
    virt_text = { { icon .. body, highlight } },
    virt_text_pos = 'inline',
    conceal = '',
]}, { 0, 1, 0, -1 })
```

Which hides the entire node and `inlines` whatever content it needs to. Basically a complete overwrite removing all the underlying information.

Compare this to the alternative:

```lua
self.marks:add(true, self.node.start_row, ctx.start_col + 2, {
    end_col = ctx.start_col + 2 + #ctx.destination + 1,
    conceal = '',
})
```

Which doesn't inline text and instead conceals a the range corresponding to the destination.

**Reply by mcDevnagh** [reply]
ah, thank you for taking the time explaining that and working with us on this.

Is there an easy way to pull the active highlight group out of the link and set the virtual text's highlight group to that? So in the above example, the virtual text can be set to render as an H1 tag

**Reply by MeanderingProgrammer** [reply]
Does the early return nil suggestion not work? i.e.

```lua
require('render-markdown').setup({
    link = {
        wiki = {
            body = function(ctx)
                if ctx.alias then
                    return nil
                end
                local diagnostics = vim.diagnostic.get(ctx.buf, {
                    lnum = ctx.row,
                    severity = vim.diagnostic.severity.HINT,
                })
                for _, diagnostic in ipairs(diagnostics) do
                    if diagnostic.source == 'zk' then
                        return diagnostic.message
                    end
                end
                return nil
            end,
        },
    },
})
```

**Reply by MeanderingProgrammer** [reply]
Or is this a different feature and you want to be able to control the highlight group of the rendered text in general

**Reply by mcDevnagh** [reply]
yes that does work. I'm sorry I wasn't clear. My example included `# [[b48069e7]]`, which pulls its virtual text from the diagnostic. That text is still using the WikiLink highlight group instead of the underlying highlight group. I understand why from your explanation. I was asking if we could grab the highlight group from the underlying text (in this case `b48069e7`) and apply it to the virtual text (in this case `Random Encounters`) without using aliasing, so the link remains `# [[b48069e7]]` and not `# [[b48069e7|Random Encounters]]` in the file itself

**Reply by MeanderingProgrammer** [reply]
I'm not entirely sure I understand what you're saying, surely you'd want to hide `b48069e7` and show `Random Encounters`. In either case this should let you do what you're looking for: https://github.com/MeanderingProgrammer/render-markdown.nvim/commit/a53ac54cebaad8dff37015d9b4c2d3b8c7d122ae.

`body` can now return either a `string` like before, or a `table` with 2 values, one for `text` and the second for the highlight (or list of highlights), i.e.: `return { 'Random Encounters', 'MyHighlightGroup }`.

**Comment by ryan-ph** [comment]
Is it possible to render the wiki link with the note's title without diagnostics? Wasn't sure about what exactly `ctx` can contain, but would appreciate if someone could link me some documentation on `ctx` and the arguments to the `body` function and I can give it a shot.
